src/modules/audio/spatializer.h

```diff
@@ -2,14 +2,46 @@
 #include "core/maf.h"
 
 typedef struct {
-  bool (*init)(void);
+	int maxSourcesHint;
+	int fixedBuffer;
+	int sampleRate;
+} SpatializerConfigIn;
+
+typedef struct {
+	bool needFixedBuffer;
+} SpatializerConfigOut;
+
+typedef struct {
+  // return true on success
+  bool (*init)(SpatializerConfigIn configIn, SpatializerConfigOut *configOut);
   void (*destroy)(void);
-  void (*apply)(Source* source, mat4 transform, const float* input /*mono*/, float* output/*stereo*/, uint32_t frames);
+
+  // input is mono, output is interleaved stereo, framesIn is mono frames, framesOut is stereo frames.
+  // Safe to assume framesIn == framesOut unless spatializer requests needFixedBuffer.
+  // Return value is number of samples written into output.
+  uint32_t (*apply)(Source* source, const float* input, float* output, uint32_t framesIn, uint32_t framesOut);
+  // called at end of frame for any "additional noise", like echo.
+  // output is stereo, frames is stereo frames, scratch is a buffer the length of output (in case that helps)
+  // return value is number of stereo frames written.
+  uint32_t (*tail)(float *scratch, float* output, uint32_t frames);
```

Comment:
Forgot to mention this in the commit note but-- apply() may now return 0 to silence a source totally